File: algorithms/swift/sorting/QuickSort.swift
import Foundation

func quickSort<T: Comparable>(_ array: [T]) -> [T] {
    guard array.count > 1 else { return array }
    
    let pivot = array[array.count / 2]
    let less = array.filter { $0 < pivot }
    let equal = array.filter { $0 == pivot }
    let greater = array.filter { $0 > pivot }
    
    return quickSort(less) + equal + quickSort(greater)
}

let array = [10, 7, 8, 9, 1, 5]
print("Original array: \(array)")
let sortedArray = quickSort(array)
print("Sorted array: \(sortedArray)")

File: algorithms/perl/sorting/quick_sort.pl
#!/usr/bin/perl
use strict;
use warnings;

sub quick_sort {
    my @arr = @_;
    return @arr if @arr < 2;
    
    my $pivot = splice @arr, int(rand(@arr)), 1;
    my @lower = grep { $_ < $pivot } @arr;
    my @higher = grep { $_ >= $pivot } @arr;
    
    return (quick_sort(@lower), $pivot, quick_sort(@higher));
}

my @arr = (10, 7, 8, 9, 1, 5);
print "Original array: " . join(", ", @arr) . "\n";
my @sorted_arr = quick_sort(@arr);
print "Sorted array: " . join(", ", @sorted_arr) . "\n";
